### ISSUE TITLE
Podcast sorting issues due to lack of support for ID3v2's TDAT frame

### ISSUE THREAD

**Issue by retrography** [ejurgensen/forked-daapd#882]
I have just noticed that FDAAPD only relies on the TYER frame to attribute a date to each podcast episode. It then appends the date with MMDD values of 0101. This means the sort order for the episodes within each year is lost. 

This looks like a simple omission, but solving  it is an important step towards full support for podcasts. Fixing it must not be very complicated either. All FDAAPD has to do is to parse the TDAT frame as well, which contains the needed MMDD values.

**Comment by retrography** [comment]
Upon further investigation, it seems that FDAAPD does actually try to parse TDA (id3v2.2) and TDAT (id3v2.3) frames. I am not a C programmer, but it seems to me that FDAAPD tries to parse these frames as full date strings, while by definition they should only contain MMDD.

Here is the relevant code:

https://github.com/ejurgensen/forked-daapd/blob/e608b763efcf84c2d5f496ae28e0de5db082f4ec/src/library/filescanner_ffmpeg.c#L238-L239

And then: 

https://github.com/ejurgensen/forked-daapd/blob/e608b763efcf84c2d5f496ae28e0de5db082f4ec/src/library/filescanner_ffmpeg.c#L96-L129

**Comment by ejurgensen** [comment]
Yes, you are right, and if it is easy to fix it should be done. But as you say it means that forked-daapd has to read it with the year field. Not sure if that is easy or not.

As far as I can tell, TDAT is a silly format that was also removed with id3v2.4 which is from 2000. So I don't think I will go to great lengths to support it.

**Comment by retrography** [comment]
Looks like it only requires minimal modification (concatenate TYER and TDAT when TDAT exists, and then parse the date as a full date string). 

I know about v2.4, but consider this: Most tagging utilities and libraries either do not support v2.4 at all, or do not use it by default, or only have partial support for it. The result is that most libraries are not v2.4-tagged.

**Comment by retrography** [comment]
I updated some tracks to id3v2.4 and used TDRL frame to put in the complete release timestamp. FDAAPD still doesn't recognize the dates. Actually, I don't think this one is even supported by FDAAPD because I can't find it in the source code. So, there doesn't seem to be any way to inform DAAPD about the actual date/time of a podcast using id3 tags.

**Comment by ejurgensen** [comment]
The reason you can't find TDRL is that forked-daapd uses ffmpeg to parse the metadata. So ffmpeg will translate the TDRL tag to a date, which forked-daapd [then will use](https://github.com/ejurgensen/forked-daapd/blob/master/src/library/filescanner_ffmpeg.c#L161). As far I can tell ffmpeg [does support TDRL](https://git.ffmpeg.org/gitweb/ffmpeg.git/blob_plain/HEAD:/libavformat/id3v2.c).

So the tags you see in forked-daapd's souce are only those that are not parsed by ffmpeg. And indeed, it looks like ffmpeg does not parse the TDAT/TYER combo. I looked at what it would take to do it in filescanner_ffmpeg.c, and it looks easy enough, so I will add it one of these days.

**Comment by retrography** [comment]
That's strange. I just rebuilt my whole metadata index and still FDAAPD doesn't show the dates correctly regardless of the presence (or absence) of TDRL. Maybe I wait until you patch the TDAT code to see if that makes a difference. Maybe we are actually looking at two different issues (one that the dates are sometimes parsed badly, and the other that the dates are not used after parsing). Once you ping us on this thread with your commit, I will rebuild the software and share the results of the test with id3v2.2, v2.3 and v2.4.

Thanks again for the great work.

**Comment by ejurgensen** [comment]
Can you run ffprobe on one of the files with TDRL? Does it output a date field?

**Comment by retrography** [comment]
You are right, ffmpeg is the issue. Below are the outputs.

mid3v2:

```
IDv2 tag info for /media/music/Podcasts/The Entrepreneurs/Eureka 178: Waterdrop.mp3
APIC=cover front,  (image/jpeg, 39215 bytes)
COMM==eng=PREROLL_1;POSTROLL_1;
TALB=The Entrepreneurs
TCON=Podcast
TCOP=Monocle
TDRC=2019
TDRL=2019-12-13
TIT2=Eureka 178: Waterdrop
TPE1=Monocle
TPE2=Monocle
```

stagger:

```
TIT2('Eureka 178: Waterdrop')
TPE1('Monocle')
TALB('The Entrepreneurs')
TDRC('2019')
TPE2('Monocle')
TCOP('Monocle')
TCON('Podcast')
TDRL('2019-12-13')
COMM(encoding='utf-8', lang='eng', desc='', text='PREROLL_1;POSTROLL_1;')
APIC(type='Front Cover(3)', desc='', mime='image/jpeg': 39215 bytes of jpeg data)
```

ffprobe:

```
Input #0, mp3, from '/media/music/Podcasts/The Entrepreneurs/Eureka 178: Waterdrop.mp3':
  Metadata:
    album           : The Entrepreneurs
    artist          : Monocle
    album_artist    : Monocle
    comment         : PREROLL_1;POSTROLL_1;
    copyright       : Monocle
    genre           : Podcast
    title           : Eureka 178: Waterdrop
    date            : 2019
  Duration: 00:13:32.17, start: 0.000000, bitrate: 128 kb/s
    Stream #0:0: Audio: mp3, 44100 Hz, stereo, s16p, 128 kb/s
    Stream #0:1: Video: mjpeg, yuvj444p(pc, bt470bg/unknown/unknown), 600x600 [SAR 72:72 DAR 1:1], 90k tbr, 90k tbn, 90k tbc
    Metadata:
      comment         : Cover (front)
```

**Comment by retrography** [comment]
For id3v2.3, see the outputs here.

mid3v2:

```
IDv2 tag info for /media/music/Podcasts/Music that Matters/Music That Matters, Vol. 701- New Year, New Me!.mp3
APIC=cover front,  (image/jpeg, 42821 bytes)
TALB=Music that Matters
TCON=Podcast
TCOP=KEXP
TDAT=1227
TIT2=Music That Matters, Vol. 701- New Year, New Me!
TPE1=KEXP
TPE2=KEXP
TYER=2019
```

stagger:

```
/media/music/Podcasts/Music that Matters/Music That Matters, Vol. 701- New Year, New Me!.mp3: ID3v2.3 tag with 9 frames
             Title: Music That Matters, Vol. 701- New Year, New Me!
            Artist: KEXP
              Date: 2019-12-27
      Album-Artist: KEXP
             Album: Music that Matters
             Genre: Podcast
           Picture: Front Cover(3)::<42821 bytes of jpeg data>
```

ffprobe:

```
Input #0, mp3, from '/media/music/Podcasts/Music that Matters/Music That Matters, Vol. 701- New Year, New Me!.mp3':
  Metadata:
    title           : Music That Matters, Vol. 701- New Year, New Me!
    artist          : KEXP
    album           : Music that Matters
    genre           : Podcast
    copyright       : KEXP
    album_artist    : KEXP
    date            : 2019-27-12
  Duration: 01:27:44.90, start: 0.000000, bitrate: 192 kb/s
    Stream #0:0: Audio: mp3, 44100 Hz, stereo, s16p, 192 kb/s
    Stream #0:1: Video: mjpeg, yuvj444p(pc, bt470bg/unknown/unknown), 600x600 [SAR 72:72 DAR 1:1], 90k tbr, 90k tbn, 90k tbc
    Metadata:
      comment         : Cover (front)
```

**Comment by ejurgensen** [comment]
I have tested with ffmpeg 3.4 and a newer one (can't remember the number), and both parsed the TDAT + TYER combo just fine, and forked-daapd stored the date from ffmpeg as a release date as it should. I also tried with id3v2.4's TDRL, and that worked as expected. So I can't really reproduce this issue. I'm thinking that what I should actually do is remove TDAT (+TDA) and TYER (+TYE) from filescanner_ffmpeg since ffmpeg seems to parse them.

That said, I did notice that for my podcasts Kid3 would list a "Release date" that was not from the above tags, and that ffprobe did not find, and that the id3v2 utility did not list. So I suppose it must have been encoded in some other way? It would of course be good if forked-daapd could read that.

**Comment by retrography** [comment]
I agree that ffmpeg already parses most sensible date tag combinations correctly, and your solution to remove TDAT and TYER from filmscanner was also the conclusion I came up with once I saw ffmpeg correctly interpreting the combination.

I also tried TDRL with several tagging suits and it is almost always detected correctly by ffmpeg. The only exception (the one I reported above) was  when writing the tags using [mp3tag](mp3tag.de). I assume the problem is in the way mp3tag writes id3v2.4 tags then.

**Comment by ejurgensen** [comment]
I have removed TDAT, TYER and related now. I found out that some of my podcasts had good dates in the TDRL field, but ffmpeg would not parse them since the id3 version id3v2.3. So they were actually tagged incorrectly, but perhaps it is common practice out there?

Kid3 was less rigid and would show them, so that's why I could see better dates there. If there was a way to make ffmpeg less rigid I would use it to get these dates, but I haven't found one.

I will close this issue since there isn't much more I can do.
